### ISSUE TITLE
Debugger Support

### ISSUE THREAD

**Issue by SuperWangKai** [munificent/wren#425]
Step over/in/out support, etc.

**Comment by munificent** [comment]
Yes, this would be fantastic. It's a lot of work, obviously, but I think it's worth looking into. I'll have to do some thinking to see how to integrate it into the VM in a way that doesn't overwhelm it with complexity or harm performance in release builds. I have some ideas.

**Comment by underscorediscovery** [comment]
I'll be posting more info on debugging soon! We're working on getting it usable and cleaning it up for a PR. It's already usable and working well, but there are things like variable introspection and such that take a bit more effort. [You can see the debugger in action here](https://imgur.com/dazexnY.mp4) through vscode.

**Comment by Geequlim** [comment]
Good news!

**Comment by fy0** [comment]
@underscorediscovery  Great! I can't wait for this!

**Comment by SuperWangKai** [comment]
How is this feature going? Any updates?

**Comment by underscorediscovery** [comment]
We're working on getting a PR up. (It's not trivial to merge upstream while Bob is occupied though, as it makes changes to the vm itself. A PR can at least be used by others in the mean time though.)

**Comment by RallyTronics** [comment]
Is this still a thing?  What's the debugger status?

**Comment by underscorediscovery** [comment]
It's definitely still a thing! I [recently](https://twitter.com/___discovery/status/1118932456063688705) got it running on Windows, fixed several bugs and updated the visual studio code plugin to the latest versions. 

The biggest blocker on it is that it was/is tied to my codebase a bit, I'm busy working on making it run against master instead (It'll be in a branch initially as it has some important changes to review). Hopefully I'll have the branch up soon.

**Comment by RallyTronics** [comment]
Great news!  I appreciate the ongoing efforts on this.

**Comment by mitack** [comment]
Do I get this right that the debugger would be VS Code based/tied ?

**Comment by ruby0x1** [comment]
The debugger support is built into the vm and isn't tied to anything specific, but the vscode plugin exposes it and makes it usable from there. If you had another system that supports debugger integration it should be possible to make it work there too.

**Comment by ruby0x1** [comment]
Just an update on where this is at since I've been working on it:

> The biggest blocker on it is that it was/is tied to my codebase a bit

I've been resolving this part here recently. My engine was running an older version of wren (for various reasons) which contributed to the disparity, blocking a PR. Pulling just the debugger parts out proved tricky considering the delta with the wren main branch and how much had changed. **I've finally updated it to latest Wren**, and I'm testing the debugger again while I continue to clean up the debugger integration. The less surface area the debugger touches the easier to maintain and keep being an experimental flag. 

This brings the PR for initial debugging support closer.

**Comment by gkjpettet** [comment]
@ruby0x1 How is progress going on this? I'm considering writing a dedicated IDE for Wren but am hesitant to start without debugger support.

**Comment by yuppox** [comment]
+1 for this feature.

**Comment by Vinschni** [comment]
I would love to have the debugger @ruby0x1 :-)

**Comment by maks** [comment]
@ruby0x1 any chance you could publish the branch with your work on the debugger? as I couldn't find anything in your fork repo (https://github.com/ruby0x1/wren) 

Even not complete or fully working code would be **very** helpful for me as I'm doing research and trying to find good examples of how a debugger can be implemented in a interpreter bytecode vm as I would like to learn how to implement that functionality for another language and implementing debuggers is one thing that's not covered in Bob's book.

**Comment by ruby0x1** [comment]
Alright everyone step one is finally here: getting the code into the repo branches and untangled 🎉

- The very wip debugger branch:
  - https://github.com/wren-lang/wren/tree/debugger
  - read this https://github.com/wren-lang/wren/blob/debugger/doc/notes/debugger.md
- The rough wip vscode extension for using the debugger:
  - https://github.com/wren-lang/wren-vscode 

This debugger [works](https://i.imgur.com/dazexnY.mp4) but has some places where it'll access a null, and has lots of needed work on things like displaying the contents of variables etc properly. Sometimes the locals can be miscalculated? It's a work in progress!

The original implementation was done by @KeyMaster- ([twitter](https://twitter.com/keymaster_)) as a learning 
experience in [luxe](https://luxeengine.com), so thanks to them!

**Comment by mhermier** [comment]
Took a quick pick at the implementation, and I don't like it as this. It
is, too much tight coupled to the host functions, seems to force network
debugging, and does not allow self hosted debugging.

**Comment by urbanawakening** [comment]
Picking at this as well.
